contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -444,34 +444,122 @@ Alternatively, deployment could be simplified by having all components (includin
-    // sensitive information to pass to the CSI driver during NodePublish.
+    // Optional: ControllerPublishSecretRef is a reference to the secret object
+    // containing sensitive information to pass to the CSI driver during
+    // ControllerPublishVolume and ControllerUnpublishVolume calls.
     // This may be empty if no secret is required. If the secret object contains
     // more than one secret, all secrets are passed.
     // +optional
-    MountSecretRef *SecretReference `json:"mountSecretRef,omitempty" protobuf:"bytes,3,opt,name=mountSecretRef"`
+    ControllerPublishSecretRef *SecretReference `json:"controllerPublishSecretRef,omitempty" protobuf:"bytes,4,opt,name=controllerPublishSecretRef"`
+}
+```
+
+If specified, the CSI external-attacher will fetch the Kubernetes secret referenced by `ControllerPublishSecretRef` and pass it to:
+1. The CSI `ControllerPublishVolume` in the `controller_publish_credentials` field.
+2. The CSI `ControllerUnpublishVolume` in the `controller_unpublish_credentials` field.
+
+For the beta implementation of this feature instead of a new filed in the `CSIPersistentVolumeSource`, the secret reference will be passed via an annotation in the PV object:
+
+```yaml
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: my-pv
+  annotations:
+    "volume.beta.kubernetes.io/controller-publish-secret-name": "mysecret"
+    "volume.beta.kubernetes.io/controller-publish-secret-namespace": "mysecretnamespace"
+spec:
+  capacity:
+    storage: 5Gi
+  accessModes:
+    - ReadWriteOnce
+  csi:
+    driver: com.example.team.csi-driver
+    volumeHandle: volumeName
+    readOnly: false
+
+```
+
+It is assumed that since `PersistentVolume` objects are non-namespaced and `CSIPersistentVolumeSource` can only be referenced via a `PersistentVolume`, only trusted users (e.g. cluster administrators) should be able to create a `PersistentVolume` objects and, thus, specify which secret to fetch.
+
+Permissions for the corresponding external-attacher may be limited to the specified namespace to prevent a compromised attacher from gaining any other secrets.
 
-    // Optional: AttachSecretRef is a reference to the secret object containing
-    // sensitive information to pass to the CSI driver during ControllerPublish.
+#### NodePublishVolume/NodeUnpublishVolume Credentials
+
+```go
+type PersistentVolumeClaimSpec struct {
+
+    // Optional: NodePublishSecretRef is a reference to the secret object
+    // containing sensitive information to pass to the CSI driver during
+    // NodePublishVolume and NodeUnpublishVolume calls.
+    // The namespace of the secret is assumed to be the same as the namespace as
+    // the pod referencing this PVC.
     // This may be empty if no secret is required. If the secret object contains
     // more than one secret, all secrets are passed.
     // +optional
-    AttachSecretRef *SecretReference `json:"attachSecretRef,omitempty" protobuf:"bytes,4,opt,name=attachSecretRef"`
```

Comment:
Previously, this design distinguished between mount secrets (made accessible to kubelets) and attach secrets (only used by the attach/detach controller). Is that distinction being removed? Under what circumstances (if any) are kubelets expected to have access to "publish secrets"?

Reply:
That distinction still exists, but the secrets are renamed to align with the name of the CSI calls:
1. `csiProvisionerSecretName` and `csiProvisionerSecretNamespace` in `StorageClass`
    * Used for:
      * CreateVolume -> Provision (called by external-provisioner)
      * DeleteVolume -> Deprovision (called by external-provisioner)
2. `ControllerPublishSecretRef *SecretReference` in `CSIPersistentVolumeSource`
    * Used for:
      * ControllerPublishVolume -> Attach (called by external-attacher)
      * ControllerUnpublishVolume -> Dettach (called by external-attacher)
3. `NodePublishSecretRef *LocalObjectReference` in `PersistentVolumeClaimSpec`
    * Used for:
      * NodePublishVolume -> Mount (Setup) (called by kubelet)
      * NodeUnpublishVolume -> Unmount (Teardown) (called by kubelet)

Kubelet is only expected to have access to 3. Not to secrets corresponding to 1 and 2.